### PR TITLE
Countries, cities and states appearing as drop down

### LINKED ISSUE #454: Countries should appear as dropdowns in volunteer search and sign up forms
Right now the users have to manually type the country in the fields. It'd be great if users can select country through a drop down list. In that case, even validation won't be required. Also, this is implemented already in systers/portal which vms is going to be a part of (correct me if i'm wrong :smile:)

more of an enhancement, IMO.

### PR DESCRIPTION
**Pull request by anjali-dhanuka**
### Description
Django cities light package implemented thus countries, cities and states appearing as drop down.

Fixes #454 

### Type of Change:
- Code

**Code/Quality Assurance Only**
- New feature (non-breaking change which adds functionality pre-approved by mentors)

How Has This Been Tested?
image
image
image

### Checklist:
- [x] My PR follows the style guidelines of this project
- [x] I have performed a self-review of my own code or materials
- [x] I have commented my code or provided relevant documentation, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] Any dependent changes have been merged 

**Code/Quality Assurance Only**
- [x] My changes generate no new warnings 
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been published in downstream modules

### CODE REVIEW

File: vms/registration/views.py [hunk 1]
```diff
@@ -140,9 +167,11 @@ def post(self, request):
 
                     vol_country = request.POST.get('vol-country')
                     vol_phone = request.POST.get('vol-phone_number')
+                    print("anjali")
```
Comment (abhishekarya286) [inline vms/registration/views.py]: Remove `print("anjali")`

File: vms/registration/views.py [hunk 2]
```diff
@@ -140,9 +167,11 @@ def post(self, request):
...
                     if (vol_country and vol_phone):
                         if not validate_phone(vol_country, vol_phone):
                             self.phone_error = True
+                            print(self.phone_error)
```
Comment (abhishekarya286) [inline vms/registration/views.py]: Ditto

File: vms/shift/models.py [hunk 3]
```diff
@@ -23,32 +24,10 @@ class Shift(models.Model):
...
-    country = models.CharField(
-        max_length=75,
-        validators=[
-            RegexValidator(r'^[(A-Z)|(a-z)|(\s)|(\-)|(\')]+$', ),
-        ],
-        blank=True,
-        null=True,
-    )
-
+    )
```
Comment (abhishekarya286) [inline vms/shift/models.py]: Please correct this ^M thing. It has something to deal with the line ending character.

File: vms/volunteer/services.py [hunk 4]
```diff
@@ -6,7 +6,11 @@
 from organization.services import (get_organization_by_name,
                                    get_organizations_ordered_by_name)
 from volunteer.models import Volunteer
+from volunteer.models import City
 
+def get_cities():
```
Comment (abhishekarya286) [inline vms/volunteer/services.py]: In this function, don't you want to return anything?

**Comment by abhishekarya286** [comment]
Apart from this, please look at the comments by codacy and try to add the functionality for automatically updating the states and cities when the country has been selected as discussed yesterday.

**Comment by abhishekarya286** [comment]
Hi @kamsuri @ayushgarg1804. Please review the PR.

**Comment by abhishekarya286** [comment]
Hi Anjali, Any updates on this? Please try to complete this ASAP.

**Comment by anjali-dhanuka** [comment]
Actually, I was thinking to implement https://github.com/yourlabs/django-autocomplete-light. This can help us in implementing the fuzzy search.

**Comment by abhishekarya286** [comment]
@anjali, go ahead with that. **LGTM**

File: vms/vms/static/vms/js/load_state.js [hunk 5]
```diff
@@ -0,0 +1,15 @@
+$(document).ready(function(){
+$("#select_country").change(function(){
+   var countryId = $(this).val(); 
+   console.log(countryId);
+  $.ajax({
+  	url: State_Url,
+  	data: {
+          'country': countryId
```
[the remark this answers is not in the document]
Comment (anjali-dhanuka) [reply vms/vms/static/vms/js/load_state.js]: This is not a string

File: vms/registration/views.py [hunk 6]
```diff
@@ -3,21 +3,25 @@
...
-from administrator.models import Administrator 
+from administrator.models import Administrator
 from organization.services import (get_organizations_ordered_by_name,
                                    get_organization_by_id)
 from registration.forms import UserForm
 from registration.phone_validate import validate_phone
 from registration.utils import volunteer_denied
 from volunteer.forms import VolunteerForm
 from volunteer.validation import validate_file
+from volunteer.models import Volunteer
```
[the remark this answers is not in the document]
Comment (anjali-dhanuka) [reply vms/registration/views.py]: It is not added in this PR

File: vms/vms/static/vms/js/load_state.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (Monal5031) [reply vms/vms/static/vms/js/load_state.js]: Yeah but JSON format uses double quotes as per standards.

File: .travis.yml [hunk 7]
```diff
@@ -29,10 +29,10 @@ before_script:
 
 script:
   - cd vms
-  - python manage.py makemigrations auth volunteer administrator organization event job shift registration
-  - python manage.py migrate --noinput --traceback --settings=vms.settings
-  - coverage run --source='.' manage.py test -v 2
-  - coverage report -m
```
Comment (Monal5031) [inline .travis.yml]: Why remove this? It is actually helpful as it gives report of the lines missed in tests alongside the corresponding file even if the coveralls fails to send a report.

File: vms/vms/static/vms/js/load_state.js [hunk 8]
```diff
@@ -0,0 +1,14 @@
+$(document).ready(function(){
+$("#select_country").change(function(){
+   var countryId = $(this).val(); 
+  $.ajax({
+  	url: StateUrl,
```
[the remark this answers is not in the document]
Comment (anjali-dhanuka) [reply vms/vms/static/vms/js/load_state.js]: It is defined in base.html

File: vms/vms/static/vms/js/load_city.js [hunk 9]
```diff
@@ -0,0 +1,16 @@
+$(document).ready(function(){
+$("#select_state").change(function(){
+ var countryId = $("#select_country").val(); 
+ var stateId = $(this).val();
+  $.ajax({
+  	url: CityUrl,
```
[the remark this answers is not in the document]
Comment (anjali-dhanuka) [reply vms/vms/static/vms/js/load_city.js]: defined in base html file

File: .travis.yml (hunk 7 above)
Comment (anjali-dhanuka) [reply .travis.yml]: I haven't removed it. Lemme check

**Comment by anjali-dhanuka** [comment]
Hi @abhishekarya286 @ayushgarg1804 @kamsuri Please review. I have added dependency in the list too i.e states filter on the basis of country and cities on the basis of states.

File: vms/vms/static/vms/js/load_city.js [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+    $("#select_state").change(function() {
+        var countryId = $("#select_country").val();
+        var stateId = $(this).val();
+        $.ajax({
+            url: CityUrl,
+            data: {
+                "country": countryId,
+                "state": stateId
+            },
+            success: function(cities) {
```
[the remark this answers is not in the document]
Comment (anjali-dhanuka) [reply vms/vms/static/vms/js/load_city.js]: @abhishekarya286 @ayushgarg1804 @ExactlyMay @Monal5031 @kamsuri I haven't seen this being implemented anywhere, is this a genuine issue? Please clarify. Should I replace success: function() with success()?

File: vms/registration/views.py [hunk 11]
```diff
@@ -47,11 +50,14 @@ def get(self, request):
...
                 'registered': self.registered,
                 'phone_error': self.phone_error,
-                'organization_list': self.organization_list
+                'organization_list': self.organization_list,
+                'country_list': self.country_list,
             })
 
     def post(self, request):
         organization_list = get_organizations_ordered_by_name()
+        country_list = Country.objects.all()
```
[the remark this answers is not in the document]
Comment (anjali-dhanuka) [reply vms/registration/views.py]: Used in rendering

File: vms/registration/views.py [hunk 12]
```diff
@@ -10,14 +10,16 @@
...
-from administrator.models import Administrator 
+from administrator.models import Administrator
 from organization.services import (get_organizations_ordered_by_name,
                                    get_organization_by_id)
 from registration.forms import UserForm
 from registration.phone_validate import validate_phone
 from registration.utils import volunteer_denied
 from volunteer.forms import VolunteerForm
 from volunteer.validation import validate_file
+from volunteer.models import Volunteer
```
[the remark this answers is not in the document]
Comment (anjali-dhanuka) [reply vms/registration/views.py]: This is not a part of this PR.

File: vms/vms/static/vms/js/load_state.js [hunk 13]
```diff
@@ -0,0 +1,14 @@
+$(document).ready(function() {
+    $("#select_country").change(function() {
+        var countryId = $(this).val();
+        $.ajax({
+            url: StateUrl,
+            data: {
+                "country": countryId
+            },
+            success: function(states) {
```
[the remark this answers is not in the document]
Comment (anjali-dhanuka) [reply vms/vms/static/vms/js/load_state.js]: Same

File: vms/vms/static/vms/js/load_state.js [hunk 14]
```diff
@@ -0,0 +1,14 @@
+$(document).ready(function() {
+    $("#select_country").change(function() {
+        var countryId = $(this).val();
+        $.ajax({
+            url: StateUrl,
```
[the remark this answers is not in the document]
Comment (anjali-dhanuka) [reply vms/vms/static/vms/js/load_state.js]: defined in base html file.

File: vms/vms/static/vms/js/load_city.js [hunk 15]
```diff
@@ -0,0 +1,16 @@
+$(document).ready(function() {
+    $("#select_state").change(function() {
+        var countryId = $("#select_country").val();
+        var stateId = $(this).val();
+        $.ajax({
+            url: CityUrl,
```
[the remark this answers is not in the document]
Comment (anjali-dhanuka) [reply vms/vms/static/vms/js/load_city.js]: same defined in base html file.

File: vms/vms/static/vms/js/load_city.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (Monal5031) [reply vms/vms/static/vms/js/load_city.js]: This issue weighs both side,
On one side if we use short-hand it becomes short, sweet and clean but again its not python.
and on the other using this current notation gives more explicit information that it is a function.

**Comment by anjali-dhanuka** [comment]
@abhishekarya286 Please review. Now if we change the country after selecting the state then the city list disappears

File: vms/registration/views.py [hunk 16]
```diff
@@ -200,7 +244,28 @@ def post(self, request):
...
+
+def load_cities(request):
+    """
+    View to return cities belonging to the selected country and state on ajax call
+
+    """
+    country_id = request.GET.get('country')
+    state_id = request.GET.get('state')
+    cities = City.objects.filter(country_id=country_id,region_id=state_id).order_by('name')
+    return render(request, 'registration/city_dropdown_list_options.html', {'cities': cities})
```
Comment (abhishekarya286) [inline vms/registration/views.py]: Hey, what if a country doesn't have states? For example, Aruba, it's a small country that has only cities. In this case, we won't be able to select cities as the city's drop-down menu will be disabled.

File: vms/event/templates/event/create.html [hunk 17]
```diff
@@ -90,74 +90,47 @@
...
-                    </div>
-                {% else %}
-                    <div class="form-group">
-                        <label class="col-md-2 control-label">{% trans "City" %}</label>
-                        <div class="col-md-8">
-                            <input class="form-control" placeholder="{% blocktrans %}Event City{% endblocktrans %}" type="text" name="city" value="{% if form.city.value %}{{ form.city.value }}{% endif %}">
-                        </div>
-                    </div>
-                {% endif %}
+            <div class="form-group">
```
Comment (abhishekarya286) [inline vms/event/templates/event/create.html]: For this page, you haven't implemented "filter based on country and state" functionality.

File: vms/event/templates/event/edit.html [hunk 18]
```diff
@@ -100,76 +100,50 @@
...
-                    </div>
-                {% else %}
-                    <div class="form-group">
-                        <label class="col-md-2 control-label">{% trans "Country" %}</label>
-                        <div class="col-md-8">
-                            <input class="form-control" placeholder="{% blocktrans %}Event Country{% endblocktrans %}" type="text" name="country" value="{% if form.country.value %}{{ form.country.value }}{% endif %}">
-                        </div>
-                    </div>
-                {% endif %}
+
```
Comment (abhishekarya286) [inline vms/event/templates/event/edit.html]: In this page, initially it should show the already selected city, country and state when the event was created.

Comment (abhishekarya286) [reply vms/event/templates/event/edit.html]: Also the filter functionality.

**Comment by abhishekarya286** [comment]
Hi @anjali-dhanuka, thank for contributing. **PTAL** at above comments.

File: vms/event/templates/event/create.html (hunk 17 above)
Comment (anjali-dhanuka) [reply vms/event/templates/event/create.html]: done

File: vms/event/templates/event/edit.html (hunk 18 above)
Comment (anjali-dhanuka) [reply vms/event/templates/event/edit.html]: done

Comment (anjali-dhanuka) [reply vms/event/templates/event/edit.html]: image

Comment (abhishekarya286) [reply vms/event/templates/event/edit.html]: Did you commit this? I can't see the changes

File: vms/event/templates/event/create.html (hunk 17 above)
Comment (abhishekarya286) [reply vms/event/templates/event/create.html]: Did you commit the changes?

File: vms/registration/views.py (hunk 16 above)
Comment (anjali-dhanuka) [reply vms/registration/views.py]: done

File: vms/vms/static/vms/js/load_state.js [hunk 19]
```diff
@@ -0,0 +1,38 @@
+/** Fetches states belonging to selected  country */
+$(document).ready(function() {
+    $("#select_country").change(function() {
+        var countryId = $(this).val();
+        $.ajax({
+            url: CheckState,
```
[the remark this answers is not in the document]
Comment (anjali-dhanuka) [reply vms/vms/static/vms/js/load_state.js]: defined in base html file

File: vms/vms/static/vms/js/load_state.js [hunk 20]
```diff
@@ -0,0 +1,38 @@
...
+                        url: CityUrl,
+                        data: {
+                            "country": countryId,
+                            "state": 0
+                        },
+                        success: function(cities) {
+                            $("#select_city").html(cities);
+                        }
+                    });
+                } else if (statecheck == 1) {
```
[the remark this answers is not in the document]
Comment (anjali-dhanuka) [reply vms/vms/static/vms/js/load_state.js]: done

File: vms/vms/static/vms/js/load_state.js [hunk 21]
```diff
@@ -0,0 +1,38 @@
...
+                    $.ajax({
+                        url: StateUrl,
```
[the remark this answers is not in the document]
Comment (anjali-dhanuka) [reply vms/vms/static/vms/js/load_state.js]: defined in base html file.

File: vms/vms/static/vms/js/load_city.js [hunk 22]
```diff
@@ -0,0 +1,17 @@
+/** Fetches cities belonging to selected state and country */
+$(document).ready(function() {
+    $("#select_state").change(function() {
+        var countryId = $("#select_country").val();
+        var stateId = $(this).val();
+        $.ajax({
+            url: CityUrl,
```
[the remark this answers is not in the document]
Comment (anjali-dhanuka) [reply vms/vms/static/vms/js/load_city.js]: defined in base html file

File: vms/vms/static/vms/js/load_state.js [hunk 23]
```diff
@@ -0,0 +1,38 @@
...
+$(document).ready(function() {
+    $("#select_country").change(function() {
+        var countryId = $(this).val();
+        $.ajax({
+            url: CheckState,
+            data: {
+                "country": countryId
+            },
+            success: function(statecheck) {
+                if (statecheck == 0) {
```
[the remark this answers is not in the document]
Comment (anjali-dhanuka) [reply vms/vms/static/vms/js/load_state.js]: done

**Comment by abhishekarya286** [comment]
Hi @anjali-dhanuka, Thanks for your contribution. There are a couple of issues that need fixing. City filter has not been implemented for event create and edit form.
 If I select a country with no state, it works fine. But then if I change the country, it doesn't change show the state drop down menu at all.
Please look into these.

File: vms/event/views.py [hunk 24]
```diff
@@ -92,12 +97,21 @@ def get_object(self, queryset=None):
         event_id = self.kwargs['event_id']
         obj = Event.objects.get(pk=event_id)
         return obj
-
+
```
[the remark this answers is not in the document]
Comment (kamsuri) [reply vms/event/views.py]: @anjali-dhanuka fix this.

**Comment by abhishekarya286** [comment]
@anjali-dhanuka Ping

**Comment by ExactlyMay** [comment]
@anjali-dhanuka is working on rebasing and fixing this PR.

File: vms/event/views.py [hunk 25]
```diff
@@ -47,17 +48,43 @@ class EventCreateView(LoginRequiredMixin, AdministratorLoginRequiredMixin,
...
+                'country_list': country_list,
             })
         else:
-            form.save()
+            event = form.save(commit=False)
+            try:
+                country_id = self.request.POST.get('country')
+                country = Country.objects.get(pk=country_id)
+                event.country = country
+            except:
```
[the remark this answers is not in the document]
Comment (Monal5031) [reply vms/event/views.py]: Use `Exception as e`

File: vms/registration/tests/test_functional_volunteer.py [hunk 26]
```diff
@@ -21,7 +21,6 @@
 from registration.tokens import account_activation_token
 from shift.utils import create_organization, create_country
 
-
```
Comment (Monal5031) [inline vms/registration/tests/test_functional_volunteer.py]: pep8 requires 2 blank line.

File: vms/registration/tests/test_functional_volunteer.py [hunk 27]
```diff
@@ -64,11 +63,12 @@ def setUpClass(cls):
         This method initiates Firefox WebDriver, WebDriverWait and
         the corresponding POM objects for this Test Class
         """
+        call_command('cities_light')
         cls.driver = webdriver.Firefox()
         cls.driver.implicitly_wait(5)
         cls.driver.maximize_window()
         cls.page = VolunteerRegistrationPage(cls.driver)
-        cls.wait = WebDriverWait(cls.driver, 10)
+        cls.wait = WebDriverWait(cls.driver, 1000)
```
Comment (Monal5031) [inline vms/registration/tests/test_functional_volunteer.py]: wait, what. why? Its in seconds. :astonished: we never go above 10

**Pull request closed without merging**